### PR TITLE
Change the type of "total*" metrics

### PR DESCRIPTION
**Pull request by manish-hashicorp**
`DBStats` exposes metrics whose definition imply it being a counter. For example, "wait_count" is defined as "total number of connections waited for" which means that during the process run the value emitted by this stat will keep increasing. This is evident by looking at the db pool source. Hence, it should be registered as a "Count" metric instead of Gauge which implies a fixed value at a point in time unrelated to previous values.

### CODE REVIEW

File: pkg/ocgorm/stats.go
Comment (morganebridges) [inline pkg/ocgorm/stats.go:121]: Count() gives the number of metric measurement points that there are, is that what we are going for?
This is an aggregation method, not a metric type just to clarify that.

I am getting that from this:
https://opencensus.io/stats/view/

If you are making the assumption that each time this metric is emitted that it is going up by one, then I think you could use it this way, essentially throwing away the value kept in the stats object and performing the metrics aggregation in datadog instead. Is that the intent?

Comment (morganebridges) [reply pkg/ocgorm/stats.go:121]: Sorry, I am new, and I haven't pulled this repo. yet, I am going to do that, I think I am missing context.

**Comment by byronwolfman** [comment]
No direct notes, but let's hold off on merging at least until Monday! I try not to be too superstitious about Friday merges, but some of the experts in this space are already on their weekend so I want to make sure they have time to review as long as this isn't an urgent change. :)

File: pkg/ocgorm/stats.go
Comment (morganebridges) [reply pkg/ocgorm/stats.go:121]: FYI, I was looking at this to try to grok what is going on, found it interesting:
https://github.com/go-gorm/prometheus/blob/master/stats.go

File: pkg/ocgorm/stats.go
Comment (Dombo) [inline pkg/ocgorm/stats.go:129]: I was initially confused/skeptical as this is a reimplementation of an opencensus contrib package which sets these aggregations. However after checking some more recently updated (opentelemetry naturally) [packages](https://github.com/XSAM/otelsql/blob/main/instruments.go#L29-L35) they also take your approach of using Gauge/Counter for select `database/sql` metrics.

Comment (Dombo) [reply pkg/ocgorm/stats.go:129]: One thing of note, there's a [common monitors TF module](https://github.com/hashicorp/terraform-datadog-cloud-common-monitors/blob/main/db_connection_wait_monitor.tf#L12) which sets a monitor on this metric which we're now changing the type of. So that likely needs to be updated as part of this going out.

**Review by Dombo — COMMENTED** [review]
Good catch, surprised it's gone unnoticed in the open source packages so long. Will approve after handling the other pieces that depend on this.

File: pkg/ocgorm/stats.go
Comment (dmalch-hashicorp) [inline pkg/ocgorm/stats.go:1]: Thank you for fixing the issue! It could explain why our connection pool metrics show conflicting diagrams. I was wondering if it would be possible to update the aggregation logic for the `SQLClientOpenConnectionsView`, `SQLClientIdleConnectionsView`, and `SQLClientActiveConnectionsView`. I think these views don't represent the current state of the connection pool.

**Review by dmalch-hashicorp — APPROVED** [review]

File: pkg/ocgorm/stats.go
Comment (byronwolfman) [reply pkg/ocgorm/stats.go:129]: This is probably a good opportunity to use TFC's new Explorer feature to track down all the workspaces where this module will need to be updated: https://app.terraform.io/app/hcp/explorer?filter=filter%5B0%5D%5Bsource%5D%5Bcontains%5D%5B0%5D%3Dcloud-common-monitors&type=modules

File: pkg/ocgorm/stats.go
Comment (morganebridges) [reply pkg/ocgorm/stats.go:121]: One additional thing I found while trying to make sense of this that I found interesting:
https://github.com/golang/go/blob/master/src/database/sql/sql.go#L1345

Looks like we have an ever-increasing value for wait count that should be reflected in the `LastValue`. Not to say this should not be a `Count`, it does make sense, but I am still confused about why the old way doesn't work.

Comment (morganebridges) [reply pkg/ocgorm/stats.go:121]: So for the gorm collection, it looks like we are registering and scheduling these OC views on an [interval](https://github.com/hashicorp/go-gin-gorm-opencensus/blob/master/pkg/ocgorm/stats.go#L167), and on that interval, we are building the DBStats object up each time. Given the definition of the OC count aggregation type: 

> The count of the number of measurement points

But we do not create a measurement point each time a connection waits, this is something that is actually set directly in the [database object](https://github.com/golang/go/blob/master/src/database/sql/sql.go#L1345) that golang/sql has to wrap the driver connection, and then is queried out when the DBStats object is [rendered](https://github.com/golang/go/blob/master/src/database/sql/sql.go#L1203).

During these scheduled events, it seems like that would be when these measurement points would be created.
If we were emitting a metric each time a connection waited, then count would make perfect sense, but since we are relying on the db object's internal reporting, I don't think it makes sense to use anything except the value that it reports.
Does this make sense?

**Comment by manish-hashicorp** [comment]
I am closing this based on https://github.com/hashicorp/go-gin-gorm-opencensus/pull/12#discussion_r1554160025. While the metric in `DBStats` is counter, `opencensus.view.Count` is not meant for that. Big thanks to @morganebridges for catching this.

**Pull request closed without merging**
